File: Classes/Base/LWAttributedStringUtils.m
//
//  LWAttributedStringUtils.m
//  LWThemeObjC
//
//  Created by lalawue on 2019/12/22.
//

#import "LWAttributedStringUtils.h"
#import "LWTheme_Tools.h"

@implementation LWAttributedStringUtils

+ (NSDictionary *)lw_mapAttrs:(NSDictionary *)attrs
{
    NSMutableDictionary *mAttrs = [NSMutableDictionary dictionaryWithDictionary:attrs];
    
    for (NSAttributedStringKey attrName in [self lw_colorKeys]) {
        NSString *colorPattern = attrs[attrName];
        if ([colorPattern isKindOfClass:[NSString class]]) {
            UIColor *colorValue = [self lw_colorForPattern:colorPattern];
            [mAttrs setObject:colorValue forKey:attrName];
        }
    }
    
    return mAttrs.copy;
}

+ (NSArray<NSAttributedStringKey> *)lw_colorKeys
{
    static NSArray *colorKeys;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        colorKeys = @[NSForegroundColorAttributeName,
                      NSBackgroundColorAttributeName,
                      NSStrokeColorAttributeName,
                      NSUnderlineColorAttributeName,
                      NSStrikethroughColorAttributeName];
    });
    return colorKeys;
}

@end

File: Classes/Base/LWAttributedStringUtils.swift
//
//  LWAttributedStringUtils.swift
//  LWThemeSwift
//
//  Created by lalawue on 2019/12/22.
//

import UIKit

class LWAttributedStringUtils {
    
    class func lw_mapAttr(_ attrs : [NSAttributedString.Key : Any]) -> [NSAttributedString.Key : Any]? {
        var mAttrs = NSMutableDictionary.init(dictionary: attrs) as! [NSAttributedString.Key : Any]
        for attrName in self.lw_colorKeys {
            let colorPattern = attrs[attrName]
            if (colorPattern is String) {
                let colorValue = LWThemeManager.sharedInstance.colorForPattern(colorPattern as? String)
                mAttrs[attrName] = colorValue
            }
        }
        return mAttrs
    }
    
    fileprivate static let lw_colorKeys = [NSAttributedString.Key.foregroundColor,
                                           NSAttributedString.Key.backgroundColor,
                                           NSAttributedString.Key.strokeColor,
                                           NSAttributedString.Key.underlineColor,
                                           NSAttributedString.Key.strikethroughColor]

}
